packages/@sanity/cli/src/actions/login/login.ts

```diff
@@ -106,77 +130,91 @@ export async function login(
+}): Promise<{token: Promise<TokenDetails>; server: Server}> {
+  const {apiHost, apiClient} = options
+  const domain = apiHost.includes('.sanity.work') ? 'www.sanity.work' : 'www.sanity.io'
+
+  const attemptPorts = callbackPorts.slice()
+  let callbackPort = attemptPorts.shift()
+
+  let resolveToken: (resolvedToken: TokenDetails | PromiseLike<TokenDetails>) => void
+  let rejectToken: (reason: Error) => void
+  const tokenPromise = new Promise<TokenDetails>((resolve, reject) => {
+    resolveToken = resolve
+    rejectToken = reject
+  })
 
-function getAuthInfo(es: EventSource): Promise<{secret: string; url: string}> {
-  const values = {secret: '', url: ''}
   return new Promise((resolve, reject) => {
-    es.addEventListener('error', (err: MessageEvent | EventWithMessage) => {
-      es.close()
-      const message = 'message' in err && typeof err.message === 'string' ? err.message : ''
-      reject(new Error(`Unable to get authorization info: ${message}`))
-    })
+    const server = http.createServer(async function onCallbackServerRequest(req, res) {
+      function failLoginRequest() {
+        res.writeHead(303, 'See Other', {Location: `https://${domain}/login/error`})
+        res.end()
+        server.close()
+      }
 
-    es.addEventListener('message', (msg) => {
-      const data = parseJson<ListenMessageData | null>(msg.data, null)
-      if (!data || !('type' in data)) {
+      const url = new URL(req.url || '/', `http://localhost:${callbackPort}`)
+      if (url.pathname !== callbackEndpoint) {
+        res.writeHead(404, 'Not Found', {'Content-Type': 'text/plain'})
+        res.write('404 Not Found')
+        res.end()
         return
       }
 
-      if (data.type === 'secret') {
-        values.secret = data.secret
-      } else if (data.type === 'url') {
-        values.url = data.url
-      } else {
+      const absoluteTokenUrl = url.searchParams.get('url')
+      if (!absoluteTokenUrl) {
+        failLoginRequest()
         return
       }
 
-      if (values.secret && values.url) {
-        resolve(values)
-      }
-    })
-  })
-}
-
-function getAuthToken(es: EventSource): Promise<string> {
-  return new Promise((resolve, reject) => {
-    es.addEventListener('success', (msg) => {
-      es.close()
-      const data = parseJson<ListenToken | null>(msg.data, null)
-      if (!data || !data.token) {
-        reject(new Error('Failed to get token from `success`-message'))
+      // https://api.sanity.io/v2021-03-25/auth/token?query=here... => /auth/token?query=here...
+      const tokenUrl = new URL(absoluteTokenUrl)
+      const tokenPathName = tokenUrl.pathname.replace(/^.*(\/auth)/, '$1')
+      const tokenPath = `${tokenPathName}${tokenUrl.search}`
+
+      let token: TokenDetails
+      try {
+        token = await apiClient({requireUser: false, requireProject: false})
+          .clone()
+          .request({uri: tokenPath})
+      } catch (err) {
+        failLoginRequest()
+        rejectToken(err)
```

Comment:
This feels a bit heavy handed given we have a fully qualified url for fetching the token. If there are additional benefits to using the `apiClient` instead of fetch, we can craft the `/auth/fetch` uri manually as one of the url parameters is the `sid`. So

```js
const sessionId = url.searchParams.get('sid')
if (sessionId) {
	failLoginRequest()
  return
}

let token: TokenDetails
try {
	token = await apiClient({requireUser: false, requireProject: false})
  	.clone()
    .request({uri: `/auth/fetch?sid=${sid}`})
} catch (err) {
	failLoginRequest()
	rejectToken(err)
}
```

Reply:
💡 good to know!